modules/core/src/main/java/org/apache/ignite/internal/processors/metric/list/MonitoringListAdapter.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.metric.list;
+
+import java.util.Iterator;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Function;
+import org.apache.ignite.spi.metric.list.MonitoringRow;
+import org.apache.ignite.spi.metric.list.MonitoringRowAttributeWalker;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Monitoring list backed by {@code data} {@link ConcurrentMap}.
+ */
+public class MonitoringListAdapter<R extends MonitoringRow, D> extends AbstractMonitoringList<R> {
+    /** Data backed by this list. */
+    private final ConcurrentMap<?, D> data;
```

Comment:
Why do we need `ConcurrentMap` here? It brings some restrictions to data source for lists. We only use `values()` and `size()` methods of `data` field, why can't we use `Collection`?

Reply:
Because Map is not a Collection, I think :)

Reply:
You use only `values()` collection from this map. I see no reason to pass map here.

And `Map` say nothing about thread safety. Thread safety provides by the implementation, not by the interface.

Reply:
> And Map say nothing about thread safety.

Thats why I use `ConcurrentMap` which declares:

```
 * A {@link java.util.Map} providing thread safety and atomicity
 * guarantees.
```

If we use `Collection` we can't force "user" of the list to pass thread-safety implementation of Collection.